Issues/7633 jetpack login

aerych opened the pull request:
Closes #7633

This PR updates the Jetpack login screen shown in the Stats feature to use the new login flow.  The approach matches the current implementation in the Android app with two differences:
- The email screen does not provide an option to use the screen name instead.
![simulator screen shot - ipad air - 2017-12-16 at 14 05 30](https://user-images.githubusercontent.com/1435271/34073934-e0b8491e-e26a-11e7-8c2c-79058fbca46e.png)

- We continue to show a modified version of the current Jetpack prompt rather than show the login flow as soon as the stats menu item is selected.   This is both for a better experience on the iPad and to retain the informative and Jetpack version checking behavior.
![simulator screen shot - ipad air - 2017-12-16 at 14 05 24](https://user-images.githubusercontent.com/1435271/34073936-e5c3de1e-e26a-11e7-92b5-ee2d0d820a54.png)

To test:
- Start signed out
- Sign in to a self-hosted site. 
- View the stats feature and tap to install Jetpack
- Progress through the flow.  Confirm you are viewing the stats screen after dismissing the login epilogue.

Test on iPhone and iPad devices. 

Bonus: For paranoia, double check the version checking and more info features of the Jetpack login screen.

@koke would you be game for this one?

koke commented:
> The email screen does not provide an option to use the screen name instead.

JetpackState contains not only the username but also the associated email address. We could use that one to at least pre-fill the email field

folletto commented:
Design wise, we are matching the login process, so no change there, right?

> We continue to show a modified version of the current Jetpack prompt rather than show the login flow as soon as the stats menu item is selected.

This will probably need an update and standardization on the design later, but I think it's a good approach for this one. 👍

koke commented:
I think this is different to the previous flow in that it will always replace the existing wp.com account.
If you log in to a different account than the one connected (easier now since it doesn't pre-fill username/email) it will "succeed" in adding the account but you'll be back at the same login prompt.

folletto commented:
> I think this is different to the previous flow in that it will always replace the existing wp.com account.

Replaces as in... would something change? If it changes in an unexpected way we should probably mention it.

koke commented:
Yes, if you're already logged it to a wp.com account and the site is connected to a different one, the login prompt is there, it will go through the login flow, and it will sign out of the previous WordPress.com account, replacing it with the new one.

folletto commented:
Ok that seems a major and unexpected change. 
We should point that out to the user somehow.

Maybe can we start with a simple dialog if the user presses the button IF there's already a logged in .com account?

> Looks like you are already logged in to WordPress.com with '$email'. Do you want to log out and log in again with a different account?
> [ Back ] [ New log in ]

Or something like this?

koke commented:
I think this is a decision we've been postponing for a long time: do we support multiple WordPress.com accounts? We never supported it officially, as there was no "account switcher", and there was always one main account to be used for reader and notifications. However, we allowed Jetpack sites to be connected to a different account for stats (and would use that account to communicate via REST instead of XML-RPC).

If we change this behavior we're preventing something that was possible before (having multiple Jetpack sites connected to different accounts). It's something we can do, but I want to make sure it's a conscious decision.

Personally, I've always leaned towards building proper multiple account support.

folletto commented:
> I think this is a decision we've been postponing for a long time: do we support multiple WordPress.com accounts? 

This seems a discussion that shouldn't happen in a PR tho. :)
I won't comment here.

One thing I'm not understanding:

> and it will sign out of the previous WordPress.com account

> we allowed Jetpack sites to be connected to a different account for stats

Can't we keep doing this then? Why do we have to log out?

aerych commented:
> I think this is different to the previous flow in that it will always replace the existing wp.com account.

Ah, right.  Its been so long since I've played around with multiple self-hosted sites + different connected Jetpack account's I'd forgot about that.  We should be able to preserve that functionality... I think.

aerych commented:
@koke I've updated the branch to restore the [previous behavior](https://github.com/wordpress-mobile/WordPress-iOS/pull/8355#issuecomment-352738409) with 6502617 .  Ptal. 

I'll confess I'm a bit skeptical about this. I did some testing in the 9.0 branch and found some very odd behavior when we do this.  Give this a try:
- Sign in to a single self-hosted account. 
- From the Me tab sign in to a wpcom account that is not connected to the self-hosted account.
- From the Stats screen, enter credentials of a different wpcom account to connect Jetpack. Notice that all the blogs associated with that wpcom account are now added to the app.
- Tour the app.  Everything at this point seems what we'd expect.
- From the Me tab, log out of wpcom.  Notice you are taken back to the login screen.  This is because there are no more "self-hosted" blogs... even tho those blogs are still present. 
- Sign back into the self-hosted account. Notice that all the wpcom blogs assoicated with the self-hosted blogs Jetpack credentials are present. Stats works but not notifications, and the Me tab shows you are logged out of wpcom.

Logging back into and out of wpcom didn't clean up the blog list, the jetpack connected blogs seemed to be orphaned.  The only way I found to clean them up was to delete and reinstall the app.

Since this behavior is present in 9.0 I'm inclined to say addressing it is out of scope for this patch but something we definitely should deal with (cc @elibud for scoping under betterjpxp).

nheagy commented on WordPress/Classes/ViewRelated/NUX/LoginEmailViewController.swift:
```diff
@@ -102,8 +102,11 @@ class LoginEmailViewController: LoginViewController, SigninKeyboardResponder {
-        instructionLabel?.text = NSLocalizedString("Log in to WordPress.com using an email address to manage all your WordPress sites.", comment: "Instruction text on the login's email addresss screen.")
-
+        if loginFields.meta.jetpackLogin {
+            instructionLabel?.text = NSLocalizedString("To view your stats, log in to the WordPress.com account you used to connect Jetpack.", comment: "Instruction text on the login's email addresss screen.")
```
`Addresss` 🐍

aerych replied:
😱

koke commented:
> Notice that all the blogs associated with that wpcom account are now added to the app.

I think they shouldn't and didn't use to. IIRC we only ever called `syncBlogs` for the default account

aerych commented:
> I think they shouldn't and didn't use to. IIRC we only ever called syncBlogs for the default account

Took a trip in the way back machine and [I see what you mean](https://github.com/wordpress-mobile/WordPress-iOS/blob/6.9/WordPress/Classes/ViewRelated/Stats/JetpackSettingsViewController.m#L453).  Looks like that [changed about a year ago](https://github.com/wordpress-mobile/WordPress-iOS/commit/dca25728be9a459874b62c1f421a8207e05ebd8b).

I'll take another look at restoring what once was.

aerych commented:
@koke, I think this is ready for another peek. 

As best I can tell, this last batch of commits restores the behavior prior to https://github.com/wordpress-mobile/WordPress-iOS/commit/dca25728be9a459874b62c1f421a8207e05ebd8b
- If there is no wpcom account already present, logging into Jetpack logs sets the wpcom account as the default wpcom account and syncs blogs.  This was existing behavior from what I could tell.
- If there is already a wpcom account present, logging into Jetpack just affects the blog in question.

When I was testing the changes the presence of the epilogue screen didn't seem to make much sense anymore since we're not replacing the default wpcom account.   I've opted to not show it and instead dismiss right into Stats.

koke commented on WordPress/Classes/Services/BlogService.m:
```diff
@@ -590,6 +590,37 @@ - (void)removeBlog:(Blog *)blog
+- (void)associateSyncedBlogsToJetpackAccount:(WPAccount *)account
+                                     success:(void (^)(void))success
+                                     failure:(void (^)(NSError *error))failure
+{
```
Just making a note here that there might be some overlap with `migrateRemoteJetpackBlog:forAccount:`.

This method is currently only associating the existing `Blog` with the new `WPAccount` if it's contained in the `me/sites` response. I'd suggest, since we just retrieved the data, that we update the matching blog with `updateBlogWithRemoteBlog:`, which in turn would call the migration.

I don't see any immediate problems if we don't and I'll keep adding on this as part of the connection flow, so feel free to skip the suggestion.

aerych replied:
I'll leave it to you then :)

koke commented:
One more thing: probably not a big deal, but if you try to log in with google with an email that's not associated to any account, it will offer the chance to add a self hosted site

![screen shot 2017-12-26 at 10 49 19](https://user-images.githubusercontent.com/8739/34353796-9be4e6cc-ea2a-11e7-85e0-5dae984b921e.png)

koke commented:
OK, this one is worse. When you already have a WordPress.com account in the app and try to authenticate the Jetpack site using a magic link, the link in the email launches the app, but you're still back ant the "Open your mail" screen

![screen shot 2017-12-26 at 10 54 38](https://user-images.githubusercontent.com/8739/34353957-541a7c5c-ea2b-11e7-8a1d-3ab82997a5c2.png)

aerych commented:
> One more thing: probably not a big deal, but if you try to log in with google with an email that's not associated to any account, it will offer the chance to add a self hosted site

We should probably hide this for wpcom only logins.  its an existing behavior so I'll open a new issue to tackle it. 

> When you already have a WordPress.com account in the app and try to authenticate the Jetpack site using a magic link, the link in the email launches the app, but you're still back ant the "Open your mail" screen

I took a look at this and I have a fix in mind. Given my imminent afk, and with everyone absent for the Holiday I think the best way to handle it is to go a head and land this PR and I'll submit a separate patch just for the magic link.  This will keep the scope smaller for a new reviewer and I can hopefully land it before I vanish. :)

aerych commented:
Thank you @koke !!!
